aten/src/ATen/native/native_functions.yaml

```diff
@@ -2727,17 +2727,18 @@
 
 - func: linear(Tensor input, Tensor weight, Tensor? bias=None) -> Tensor
   python_module: nn
+  dispatch:
+    CompositeImplicitAutograd: linear
+    MPS: _mps_linear
 
-- func: linear.out(Tensor input, Tensor weight, Tensor? bias=None, *, Tensor(a!) out) -> Tensor(a!)
-  python_module: nn
+- func: linear_backward(Tensor self, Tensor grad_output, Tensor weight, bool[3] output_mask) -> (Tensor, Tensor, Tensor)
+  dispatch:
+    MPS: mps_linear_backward
 
-# TODO: Add this function to MPS dispatch key so that we avoid declaring it in
-# native_functions.yaml
-# https://github.com/pytorch/pytorch/issues/77394
-- func: _mps_linear(Tensor self, Tensor weight, Tensor? bias=None) -> Tensor
+- func: linear.out(Tensor input, Tensor weight, Tensor? bias=None, *, Tensor(a!) out) -> Tensor(a!)
   python_module: nn
   dispatch:
-    MPS: _mps_linear
+    CompositeExplicitAutograd: linear_out
```

Comment:
@albanD 
This diff looks kind of weird, but the second change needed to work was making linear.out composite explicit autograd.  The reason the codegen assert was failing was because linear was abstract while linear.out was not abstract. Another solution would be register a `MPS: mps_linear_out` kernel for linear.out. However since this function was previously implicit and @bdhirsh  mentioned that out variants aren't completely working with Autograd it made sense to make this composite explicit.

Reply:
Sounds good! Does the codegen still generate the implementation for this though?

Reply:
I think yes

```
 m.impl("linear.out",
         TORCH_FN(VariableType::linear_out_out)
  );
 ```

Reply:
whoop sorry yeah its getting registered here
```
m.impl("linear.out",
TORCH_FN(wrapper_out_linear_out_out));
```

Reply:
Side note I have idea why CI is failing for this I am getting no codegen errors when building locally

Reply:
Ok!
For CI, you might need to clean locally?
Maybe @bdhirsh knows about this error?

Reply:
@soulitzer  is very smart and figured that it may be because most builds have USE_MPS=0. I tried this setting locally and got the same error as CI. I tried renaming the composite implicit kernel to another name and still get a codegen error.

Reply:
oof. @albanD is the codegen set up to rip out `MPS` `dispatch` entries if an mps flag isn't passed in?

That would probably cause the error check to trip, since if the codegen removes the `MPS` entry in non-mps builds then you end up with:
```
- func: linear(Tensor input, Tensor weight, Tensor? bias=None) -> Tensor
  python_module: nn
  dispatch:
    CompositeImplicitAutograd: linear
```

Reply:
yeah, happens here right before we parse the yaml :( https://github.com/pytorch/pytorch/blob/e98e7fe42811b656bd272e92044df67ccb2d3880/torchgen/gen.py#L2460

Reply:
I guess the easiest fix is just to rename the composite kernel to something different, like `linear_math`.

Reply:
```
AssertionError: unexpected name for singleton CompositeImplicitAutograd dispatch entry: expected linear but got BackendMetadata(kernel='linear_math', structured=False, cpp_namespace='at::native').  Rename your implementation to the expected name, then delete the dispatch table
```
Codegen has thought of that as well.. 🙈

Reply:
@albanD how much do you hate the idea of saving the `mps` CLI arg as a global, and skipping the check when it's not set? :(

(Or I guess we could consider getting rid of the second error check, although it's nice to enforce the extra structure on `native_functions.yaml`)

Reply:
> yeah, happens here right before we parse the yaml :(
> 
> https://github.com/pytorch/pytorch/blob/e98e7fe42811b656bd272e92044df67ccb2d3880/torchgen/gen.py#L2460

I think this must be for not building MPS code for MacOS < 12.3

Reply:
@albanD Also now that nested linear has landed the issue of the dispatchkey getting deleted is gone(at least masked) for this example.

Reply:
haha ok!

This was only to not generate MPS code when we don't need it.
cc @bdhirsh we might not want to do https://github.com/pytorch/pytorch/issues/77251 afterall?

Reply:
Will not doing #77251, and removing that logic for MPS, break MPS still? If not, then not doing that seems fine (speeding up the codegen by a bit probably won’t noticeably improved build times)